src/main/java/com/limechain/trie/structure/slab/Slab.java

```diff
@@ -16,70 +15,157 @@
 import java.util.stream.IntStream;
 import java.util.stream.Stream;
 
+/**
+ * Represents a slab data structure for efficient management of a fixed type of objects.
+ * It provides methods for adding, removing, and accessing elements, along with utilities
+ * for managing the slab's capacity and contents.
+ *
+ * @param <T> the type of elements held in this slab
+ */
 public class Slab<T> implements Iterable<Pair<Integer, T>> {
     private ArrayList<T> storage;
     private int newIndex;
     private Queue<Integer> freeIndices;
+    private int size;
 
+    /**
+     * Constructs an empty Slab with default initial capacity of 0.
+     */
     public Slab() {
         initialize(0);
     }
 
+    /**
+     * Constructs a Slab with the specified initial capacity.
+     *
+     * @param initialCapacity the initial capacity of the slab
+     */
     public Slab(int initialCapacity) {
         initialize(initialCapacity);
     }
 
+    /**
+     * Initializes the internal storage structures of the slab.
+     *
+     * @param initialCapacity the initial capacity of the slab
+     */
     private void initialize(int initialCapacity) {
         this.storage = new ArrayList<>(initialCapacity);
         this.newIndex = 0;
+        this.size = 0;
         this.freeIndices = new LinkedList<>();
     }
 
+    /**
+     * Adds an element to the slab and returns its index.
+     *
+     * @param element the element to be added, must not be null
+     * @return the index at which the element is stored
+     * @throws NullPointerException if the element is null
+     */
     public int add(@NonNull T element) {
-        int index = freeIndices.isEmpty() ? newIndex++ : freeIndices.poll();
-        storage.add(index, element);
+        int index;
+        if (freeIndices.isEmpty()) {
+            index = newIndex;
+            storage.add(element);
+            newIndex++;
+        } else {
+            index = freeIndices.poll();
+            storage.set(index, element);
+        }
+        size++;
         return index;
     }
 
+    /**
+     * Removes the element at the specified index from the slab.
+     *
+     * @param index the index of the element to be removed
+     * @return the element previously at the specified index
```

Comment:
We should consider whether we want to throw if queried for an invalid index or always return null.
Because right now:
- we return null if the index has been used once, but freed later;
- we throw an `IndexOutOfBoundsException` if the index is out of bounds for the underlying `storage`.

I'd say the stance of this data structure is rather: "I provide you with valid indices, if you try to use me with an index (int) not explicitly guaranteed to be valid by me, it's an end-of-worlds scenario and I'll throw", but we might also go the other way and simply return null on invalid indices (thus users can't insert `null` as a value, which is already set in the contract of `add()`).

Reply:
Now that I think about it, better return null if the index is invalid. Stick to the notion of "null means no value for the given index, i.e. invalid index".